src/server/routes/entity/entity.tsx

```diff
@@ -1312,3 +1341,186 @@ export function displayPreview(req:PassportRequest, res:$Response, next) {
+	res: $Response
+) {
+	const {orm}: {orm?: any} = req.app.locals;
+	const {Entity, Revision, bookshelf} = orm;
+	const editorJSON = req.user;
+
+	const {body}: {body: Record<string, any>} = req;
+	let currentEntity: {
+		aliasSet: {id: number} | null | undefined,
+		annotation: {id: number} | null | undefined,
+		bbid: string,
+		disambiguation: {id: number} | null | undefined,
+		identifierSet: {id: number} | null | undefined,
+		type: EntityTypeString
+	} | null | undefined;
+
+	const entityEditPromise = bookshelf.transaction(async (transacting) => {
+		const savedMainEntities = {};
+		// map dummy id to real bbid
+		const bbidMap = {};
+		const allRelationships = {};
+		// callback for creating entity
+		async function processEntity(entityKey:string) {
+			const entityForm = body[entityKey];
+			const entityType = _.upperFirst(entityForm.type);
+			// edition entity should be on the bottom of the list
+			if (entityType === 'Edition' && !_.isEmpty(entityForm.publishers)) {
+				entityForm.publishers = entityForm.publishers.map((id) => bbidMap[id] ?? id);
+			}
+			allRelationships[entityKey] = entityForm.relationships;
+			const newEntity = await new Entity({type: entityType}).save(null, {transacting});
+			currentEntity = newEntity.toJSON();
+
+			// create new revision for each entity
+			const newRevision = await new Revision({
+				authorId: editorJSON.id,
+				isMerge: false
+			}).save(null, {transacting});
+			const additionalProps = _.pick(entityForm, additionalEntityProps[_.snakeCase(entityType)]);
+			const changedProps = await getChangedProps(
+				orm, transacting, true, currentEntity, entityForm, entityType,
+				newRevision, additionalProps
+			);
+			const mainEntity = await fetchOrCreateMainEntity(
+				orm, transacting, true, currentEntity.bbid, entityType
+			);
+			mainEntity.shouldInsert = true;
+
+			// set changed attributes on main entity
+			_.forOwn(changedProps, (value, key) => mainEntity.set(key, value));
+			const savedMainEntity = await saveEntitiesAndFinishRevision(
+				orm, transacting, true, newRevision, mainEntity, [mainEntity],
+				editorJSON.id, entityForm.note
+			);
+
+			/* We need to load the aliases for search reindexing and refresh it*/
+			await savedMainEntity.load('aliasSet.aliases', {transacting});
+
+			/* New entities will lack some attributes like 'type' required for search indexing */
+			await savedMainEntity.refresh({transacting});
+
+			/* fetch and reindex EditionGroups that may have been created automatically by the ORM and not indexed */
+			if (savedMainEntity.get('type') === 'Edition') {
+				await indexAutoCreatedEditionGroup(orm, savedMainEntity, transacting);
+			}
+			bbidMap[entityKey] = savedMainEntity.get('bbid');
+			savedMainEntities[entityKey] = savedMainEntity.toJSON();
+		}
+		try {
+			// bookshelf's transaction have issue with Promise.All
+			await Object.keys(body).reduce((promise, entityKey) => promise.then(() => processEntity(entityKey)), Promise.resolve());
```

Comment:
This line is quite hard to read.
Can we look at fixing the underlying issue? Can you explain what the symptom was that you had when using Promise.all?
If we can't fix it, at least it would be good to explain in a comment what the code is doing

Reply:
Not sure if we can fix that, it seems to be common problem with bookshelf when using transaction with promise.all as described [here](https://github.com/bookshelf/bookshelf/issues/1498) 
I will add this issue link in the comment as well